src/Workspaces/Core/Portable/Shared/Extensions/SafeHandleLease.cs

```diff
@@ -0,0 +1,30 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.CodeAnalysis.Shared.Extensions
+{
+    /// <summary>
+    /// Represents a lease of a <see cref="SafeHandle"/>.
+    /// </summary>
+    /// <seealso cref="SafeHandleExtensions.Lease"/>
+    internal readonly struct SafeHandleLease : IDisposable
+    {
+        private readonly SafeHandle? _handle;
+
+        internal SafeHandleLease(SafeHandle handle)
+            => _handle = handle;
+
+        /// <summary>
+        /// Releases the <see cref="SafeHandle"/> lease. The behavior of this method is unspecified if called more than
+        /// once.
+        /// </summary>
+        public void Dispose()
+            => _handle?.DangerousRelease();
```

Comment:
Generally dispose is supposed to be idempotent.  Here if someone accidentally disposed of the SafeHandleLease twice, it would release the SafeHandle twice, which could cause reliability issues.

Reply:
This type is a `struct`, so without support for non-copyable we have to worry about this anyway. The use case is documented on the `Lease()` extension method as being intended to appear in a `using` block.